client/src/App.js

```diff
@@ -13,67 +21,145 @@ function App() {
+  const [loading, setLoading] = useState(false)
+
+  const [offsetCounter, setOffsetCounter] = useState(0)
+
+  const addIngredient = (ingredient) => {
     const newIngredient =
     {
-      name: ingredient.name,
-      id: ingredient.id,
+      ...ingredient,
       isActive: true
     }
-
-    setIngredients([newIngredient, ...ingredients])
-    console.log(ingredients)
+    if (isValidId(newIngredient) && isNewEntry(ingredients, newIngredient)) {
+      setIngredients([newIngredient, ...ingredients])
+    }
   }
 
   const deleteIngredient = (idToDelete) => {
-    const ingredientsToKeep = ingredients.filter(ingredient => (ingredient.id !== idToDelete))
-
+    const ingredientsToKeep = deleteItem(ingredients, idToDelete)
     setIngredients(ingredientsToKeep)
     setActiveIngredients(ingredientsToKeep)
   }
 
   const toggleActiveState = (idToToggle) => {
-    const updatedIngredients = ingredients.map(ingredient => {
-      if (ingredient.id === idToToggle) {
-        ingredient.isActive = !ingredient.isActive
-      }
-      return ingredient;
-    })
+    const updatedIngredients = toggleIngredient(ingredients, idToToggle)
     setIngredients(updatedIngredients)
   }
 
-  function filterActiveIngredients() {
-    const allActiveIngredients = ingredients.filter(ingredient => ingredient.isActive);
-    setActiveIngredients(allActiveIngredients)
+  useEffect(() => {
+    const allActiveIngredients = filterActiveIngredients(ingredients)
     saveToLocal('ingredients', ingredients)
     saveToLocal('activeIngredients', allActiveIngredients)
+  }, [ingredients])
+
+  const getRecipeResults = async () => {
+    setLoading(true)
+    const recipeData = await getRecipeData(activeIngredients, offsetCounter)
+    if (recipeData) {
+      setRecipes(recipeData)
+      saveToLocal('recipes', recipeData)
+      setLoading(false)
+      setError(false)
+    } else {
+      setLoading(false)
+      setError(true)
+    }
+  }
+
+  function deleteRecipe(idToDelete) {
+    const recipesToKeep = deleteItem(recipes, idToDelete)
+    setRecipes(recipesToKeep)
+    saveToLocal('recipes', recipesToKeep)
+  }
+
+  function addToLikedRecipes(recipeToAdd) {
+    const newRecipe = {
+      ...recipeToAdd,
+      isLiked: true
+    }
+
+    if (isNewEntry(likedRecipes, newRecipe)) {
+      setLikedRecipes([newRecipe, ...likedRecipes])
+    }
+
+    const allUnlikedRecipes = recipes.filter(recipe => recipe.id !== recipeToAdd.id);
+    setRecipes(allUnlikedRecipes)
+    saveToLocal('recipes', allUnlikedRecipes)
   }
 
   useEffect(() => {
-    filterActiveIngredients()
-  }, [ingredients])
+    saveToLocal('likedRecipes', likedRecipes)
+  }, [likedRecipes])
+
+
+  const increaseOffsetCounter = () => {
+    setOffsetCounter(offsetCounter + 6)
+    return offsetCounter
+  }
+
+  const getNextRecipeResults = async () => {
+    setLoading(true)
+    increaseOffsetCounter()
+    const nextRecipeData = await getRecipeData(activeIngredients, offsetCounter + 6)
```

Comment:
Not sure again, but you might be able to pass `increaseOffsetCounter()` instead of `offsetCounter + 6` and could reuse your function like this.

Reply:
Unfortunately I had some problems with the offsetCounter non increasing when I wanted it to because of synchronicity. I will leave it for now because it works, but I will definitely try to figure out a better way later